lnd_test.go

```diff
@@ -1876,35 +1904,295 @@ func testRevokedCloseRetribution(net *networkHarness, t *harnessTest) {
+	// update, then ensure that the closing transaction was included in the
+	// block.
+	block := mineBlocks(t, net, 1)[0]
+
+	breachTXID, err := net.WaitForChannelClose(ctxb, closeUpdates)
+	if err != nil {
+		t.Fatalf("error while waiting for channel close: %v", err)
+	}
+	assertTxInBlock(t, block, breachTXID)
 
 	// Query the mempool for Alice's justice transaction, this should be
 	// broadcast as Bob's contract breaching transaction gets confirmed
 	// above.
-	var justiceTXID *chainhash.Hash
-	breakTimeout := time.After(time.Second * 5)
-poll:
-	for {
-		select {
-		case <-breakTimeout:
-			t.Fatalf("justice tx not found in mempool")
-		default:
+	justiceTXID, err := waitForTxInMempool(net.Miner.Node, 5*time.Second)
+	if err != nil {
+		t.Fatalf("unable to find Alice's justice tx in mempool: %v", err)
+	}
+	time.Sleep(100 * time.Millisecond)
+
+	// Query for the mempool transaction found above. Then assert that all
+	// the inputs of this transaction are spending outputs generated by
+	// Bob's breach transaction above.
+	justiceTx, err := net.Miner.Node.GetRawTransaction(justiceTXID)
+	if err != nil {
+		t.Fatalf("unable to query for justice tx: %v", err)
+	}
+	for _, txIn := range justiceTx.MsgTx().TxIn {
+		if !bytes.Equal(txIn.PreviousOutPoint.Hash[:], breachTXID[:]) {
+			t.Fatalf("justice tx not spending commitment utxo "+
+				"instead is: %v", txIn.PreviousOutPoint)
 		}
+	}
+
+	// We restart Alice here to ensure that she persists her retribution state
+	// and successfully continues exacting retribution after restarting. At
+	// this point, Alice has broadcast the justice transaction, but it hasn't
+	// been confirmed yet; when Alice restarts, she should start waiting for
+	// the justice transaction to confirm again.
+	if err := net.RestartNode(net.Alice, nil); err != nil {
+		t.Fatalf("unable to restart Alice's node: %v", err)
+	}
 
-		mempool, err := net.Miner.Node.GetRawMempool()
+	// Now mine a block, this transaction should include Alice's justice
+	// transaction which was just accepted into the mempool.
+	block = mineBlocks(t, net, 1)[0]
+
+	// The block should have exactly *two* transactions, one of which is
+	// the justice transaction.
+	if len(block.Transactions) != 2 {
+		t.Fatalf("transaction wasn't mined")
+	}
+	justiceSha := block.Transactions[1].TxHash()
+	if !bytes.Equal(justiceTx.Hash()[:], justiceSha[:]) {
+		t.Fatalf("justice tx wasn't mined")
+	}
+
+	// Finally, obtain Alice's channel state, she shouldn't report any
+	// channel as she just successfully brought Bob to justice by sweeping
+	// all the channel funds.
+	req := &lnrpc.ListChannelsRequest{}
+	aliceChanInfo, err := net.Alice.ListChannels(ctxb, req)
+	if err != nil {
+		t.Fatalf("unable to query for alice's channels: %v", err)
+	}
+	if len(aliceChanInfo.Channels) != 0 {
+		t.Fatalf("alice shouldn't have a channel: %v",
+			spew.Sdump(aliceChanInfo.Channels))
+	}
+}
+
+func testRevokedCloseRetributionPostBreachConf(
```

Comment:
Or if this is to be refactored then it might not be that important, but because of the duplication atm it is hard to quickly glance where they are different.